### PR TITLE
Add MRChem harness

### PR DESCRIPTION
**Pull request by robertodr**
## Description
This adds a harness for [MRChem](mrchem.readthedocs.io), the multiresolution analysis DFT code developed at the university of Tromsø.

### TODOs
- [X] Proper handling of all drivers.
- [x] Test MPI invocation.
- [X] Proper translation between MRChem output schema and QCSchema output.
- [X] Translation from QCElemental molecule to MRChem ~~: MolSSI/QCElemental#238 pending~~

## Changelog description

Add a harness for [MRChem](mrchem.readthedocs.io), the multiresolution analysis DFT code developed at the University of Tromsø.

## Status

- [X] Code base linted
- [x] Ready to go

### CODE REVIEW

**Comment by robertodr** [comment]
How should I go about testing the harness in CI? There is no conda package for MRChem.

File: qcengine/programs/mrchem.py [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+
+        mrchemrec = {
+            "scratch_directory": config.scratch_directory,
+        }
+
+        opts = copy.deepcopy(input_model.keywords)
+
+        # Handle molecule
+        _, moldict = input_model.molecule.to_string(
+            dtype="mrchem", units=opts.get("world_unit", "bohr"), return_data=True
```
Comment (loriab) [inline qcengine/programs/mrchem.py:208]: you're probably waiting on a qcel release to be able to write a few mrchem harness tests for here. let me fix that ...

**Review by loriab — COMMENTED** [review]
looking good! if the timeline works for you, let me do a qcel release so you can test this, then this PR can merge for the following qcng release.

File: qcengine/programs/mrchem.py [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+        )
+        opts["Molecule"] = moldict["keywords"]
+
+        if "WaveFunction" in opts.keys():
+            opts["WaveFunction"]["method"] = input_model.model.method
+        else:
+            opts["WaveFunction"] = {"method": input_model.model.method}
```
[the remark this answers is not in the document]
Comment (robertodr) [reply qcengine/programs/mrchem.py:215]: SCF is HF or DFT. No other methods are implemented in MRChem.

**Comment by robertodr** [comment]
> looking good! if the timeline works for you, let me do a qcel release so you can test this, then this PR can merge for the following qcng release.

Hello! We are ironing out some details of the code to prepare its official v1.0.0 This release will have a conda package (on conda-forge) which will help with testing the harness in CI. We don't have a strict timeline for the release, but it's highly likely it will happen within the next two weeks. I'd say go ahead with the QCElemental release, but hold off a little while with merging this one. Should I set it back to "Draft"?

**Comment by loriab** [comment]
that sounds good. will do qcel and hold back on this one. no need to set it to draft. please ping when ready for review!

File: qcengine/programs/mrchem.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+                    for x, z in known_rsp_props
+                    if x in mrchem_output["properties"]
+                    for y in mrchem_output["properties"][x].keys()
+                ]
+
+                # fill up extras:
+                # * under "raw_output" the whole JSON output from MRChem
+                # * under "properties" all the properties computed by MRChem
+                output_data["extras"] = {
+                    "raw_output": mrchem_json,
```
[the remark this answers is not in the document]
Comment (robertodr) [reply qcengine/programs/mrchem.py:159]: The standard output is similar to Psi4's `output.dat`, a human-readable log of what happened during the calculation. The JSON file contains the data in a format more suitable for programmatic access. It contains both the input and the output.

File: qcengine/programs/mrchem.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+
+                # fill up return_result
+                if input_model.driver == "energy":
+                    output_data["return_result"] = mrchem_output["properties"]["scf_energy"]["E_tot"]
+                elif input_model.driver == "properties":
+                    output_data["return_result"] = {
+                        f"{ks[1]}": {f"{ks[2]}": _nested_get(mrchem_output, ks)} for ks in computed_rsp_props
+                    }
+                else:
+                    raise RuntimeError(f"MRChem cannot run with {input_model.driver} driver")
```
[the remark this answers is not in the document]
Comment (robertodr) [reply qcengine/programs/mrchem.py:173]: We might yet enable `gradient` as a valid driver :)

**Comment by robertodr** [comment]
I have added a test in `qcengine/programs/tests` and all that's necessary to run it in CI. However, until this PR conda-forge/staged-recipes#12679 gets merged, it cannot be run.

File: qcengine/programs/mrchem.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (robertodr) [reply qcengine/programs/mrchem.py]: The gradient didn't make the cut for the release. I have converted to `InputError`

**Review by loriab — APPROVED** [review]
lgtm!

ready for merge after testing in place. add yourself to codemeta, too.

**Comment by robertodr** [comment]
@loriab: testing is in place, though Travis seems to be stuck.

**Comment by robertodr** [comment]
Ready to go!

**Comment by loriab** [comment]
Travis unstuck, and I reran the MRChem lane again to make sure still good to go. Thanks!

**Pull request merged**